### ISSUE TITLE
PhantomData is incorrectly skipped when used as type param

### ISSUE THREAD

**Issue by RReverser** [eqrion/cbindgen#99]
```rust
#[repr(C)]
pub struct S<T1, T2> {
  x: T1,
  y: T2
}

#[no_mangle]
pub extern fn f2(arg: S<PhantomData<()>, i32>) {}
```

incorrectly produces

```cpp
#include <cstdint>
#include <cstdlib>

template<typename T1, typename T2>
struct S {
  T1 x;
  T2 y;
};

extern "C" {

void f2(S<int32_t> arg);

} // extern "C"
```

(note the number of params in `S<int32_t>` versus expected). Maybe has to do with recent implementation of #85 – likely was skipped before, but revealed now because we avoid monomorphisation.

**Comment by eqrion** [comment]
I agree - I think this was skipped before and now the lack of monomorphisation makes it a problem.

Zero sized types are kind of clunkily treated as not existing at all by returning `None` from the loading function. In this case, we need to be smarter.

**Comment by RReverser** [comment]
Thinking again, it probably is exactly a case where `template` approach might be breaking things (what I've been asking in https://github.com/eqrion/cbindgen/issues/85#issue-274253497). Even if zero-sized Rust type could be passed to such generic, without monomorphisation it will actually be non-zero-sized in C++, which means field offsets will become wrong.

Not sure what to do with this TBH. On one hand, this is a breaking change, on another, this representation seems much cleaner and really wouldn't want to revert it for 90% of cases in favour of this 10% where zero-sized types are passed as generics...

**Comment by RReverser** [comment]
@eqrion What are your thoughts on this zero-sized types in generics issue? Can we live with that tradeoff or should we switch to the old approach?

**Comment by eqrion** [comment]
I think it's okay to live with the tradeoff. In `webrender` the biggest use case of zero sized types is as a parameter to a `PhantomData` which itself is zero sized so it is properly ignored by the template approach. For example:

```rust
#[repr(C)]
struct Vector2D<Unit> {
  _unit: PhantomData<Unit>,
}
```

But I also think this means we should keep the existing code for monomorphs, and possibly enable it as an option for C++. I'm not set on this though, and if it'd simplify the code a lot it could be worth it.
